### PR TITLE
action: move and inline LogStart() further up

### PR DESCRIPTION
**Pull request by evelikov**
Currently each action starts with LogStart(), adding some annoying duplication. In addition we have a LogStart(), whereas there's no end/stop equivalent.

Instead add the on-line log.Print() wihtin the debos action loop, just prior to calling Run(). The same file deals with all the other pretty logging.

NOTE: this conflicts trivially with https://github.com/go-debos/debos/pull/377

### CODE REVIEW

File: cmd/debos/debos.go
Comment (obbardc) [inline cmd/debos/debos.go:31]: does this also print when running nest recipes with the `recipe` action? :p

Comment (evelikov) [reply cmd/debos/debos.go:31]: Oops fixed, or at least `tests/recipes/test.yaml` seems to be happy.

File: action.go
Comment (obbardc) [inline action.go:78]: to remove duplicates of Printf, I think it'd be best to keep this function and call it from the the main function in the two places ?

File: action.go
Comment (obbardc) [inline action.go:6]: keep me plz

File: actions/apt_action.go
Comment (obbardc) [inline actions/apt_action.go:46]: missing for arch actions ;-)

File: action.go
Comment (evelikov) [reply action.go:78]: Not sure I follow: are you saying that using two "log.Printf(...)" lines is too much repetition?

In general it make sense to keep some form of abstraction assuming a) it's more than one-liner b) a specific format is used and/or c) there are symmetric LogEnd/LogFoo helpers. Without any of these the function seems like a unnecessary distraction IMHO.

File: action.go
Comment (evelikov) [reply action.go:6]: If I keep it (yet remove LogStart as per above) the build will fail, since it's unused. Am I missing something?

File: actions/apt_action.go
Comment (evelikov) [reply actions/apt_action.go:46]: Correct - there is a big NOTE that aims to highlight that

File: action.go
Comment (obbardc) [reply action.go:78]: Yeah I think two separate instances is too much repetition for something like this where the same behaviour is in two places - if the style of one printf call is modified, the other call (just for nested recipes) may get missed. I'd prefer to move this function to `cmd/main/debos.go`.

File: action.go
Comment (obbardc) [reply action.go:6]: This refers to https://github.com/go-debos/debos/pull/380#discussion_r1060542966

File: action.go
Comment (evelikov) [reply action.go:78]: Shouldn't there be a CI test to catch such changes/issues? Would it make sense to add "LogEnd" in that case?

File: action.go
Comment (evelikov) [reply action.go:6]: I suspect you meant the import should be kept alongside the helper, right?

The linked suggests moving the helper to another file, yet this comment says the import should be kept as-is, which will lead to build failure.

Comment (obbardc) [reply action.go:6]: yes, lets just resolve this one instead...

File: action.go
Comment (obbardc) [reply action.go:78]: My opinion is no to both of those questions.

File: actions/apt_action.go
Comment (obbardc) [reply actions/apt_action.go:46]: right I only looked at the commit description & contents.

File: action.go
Comment (evelikov) [reply action.go:78]: Your suggestion does not seem to be doable. For example, applying the following simple diff fails to build as below.

```diff
index 9f390db..0c2e426 100644
--- a/actions/pacman_action.go
+++ b/actions/pacman_action.go
@@ -25,8 +25,6 @@ type PacmanAction struct {
 }

 func (p *PacmanAction) Run(context *debos.DebosContext) error {
-       p.LogStart()
-
        pacmanOptions := []string{"pacman", "-Syu", "--noconfirm"}
        pacmanOptions = append(pacmanOptions, p.Packages...)

diff --git a/actions/pacstrap_action.go b/actions/pacstrap_action.go
index 25c1624..e0eb9d2 100644
--- a/actions/pacstrap_action.go
+++ b/actions/pacstrap_action.go
@@ -84,8 +84,6 @@ func (d *PacstrapAction) PreMachine(context *debos.DebosContext, m *fakemachine.
 }

 func (d *PacstrapAction) Run(context *debos.DebosContext) error {
-       d.LogStart()
-
        files := map[string]string{
                "/etc/pacman.conf":         d.Config,
                "/etc/pacman.d/mirrorlist": d.Mirror,
diff --git a/cmd/debos/debos.go b/cmd/debos/debos.go
index 951cd2e..513b3b7 100644
--- a/cmd/debos/debos.go
+++ b/cmd/debos/debos.go
@@ -15,6 +15,10 @@ import (
        "github.com/jessevdk/go-flags"
 )

+func (b *debos.BaseAction) LogStart() {
+       log.Printf("==== %s ====\n", b)
+}
+
 func checkError(context *debos.DebosContext, err error, a debos.Action, stage string) int {
        if err == nil {
                return 0
```

```shell
$ go build -v ./...
github.com/go-debos/debos/cmd/debos
# github.com/go-debos/debos/cmd/debos
cmd/debos/debos.go:18:7: cannot define new methods on non-local type debos.BaseAction
````

Comment (obbardc) [reply action.go:78]: > Your suggestion does not seem to be doable. For example, applying the following simple diff fails to build as below.

Maybe you missed my edit to the suggestion:
> I'd prefer to keep this function and call it from `cmd/main/debos.go` in the two places where it is used.

Comment (obbardc) [reply action.go:78]: @evelikov I implemented the above suggestion at https://github.com/go-debos/debos/pull/417 and will close this PR. Can you review it?

**Comment by evelikov** [comment]
> @evelikov Can you please rebase onto latest main? I tried myself but cannot push to your fork.

Do you have maintainer access? The `Allow edits and access to secrets by maintainers` is ticked.

Either way - rebase pushed. Tests are failing badly here (even against main), so it'll leave it to Github to check if I haven't broken anything.

**Comment by obbardc** [comment]
> > @evelikov Can you please rebase onto latest main? I tried myself but cannot push to your fork.
> Do you have maintainer access? The `Allow edits and access to secrets by maintainers` is ticked.

Yes - as spoken about many times I am a maintainer of this project ;-)

> Either way - rebase pushed. Tests are failing badly here (even against main), so it'll leave it to Github to check if I haven't broken anything.
Thanks - curious to see which tests pass as GitHub says `All checks have passed - 23 successful and 2 skipped checks`

**Review by obbardc — APPROVED** [review]

**Comment by obbardc** [comment]
> In case you're curious here's the go test log

Right, seems like you're just missing `dpkg` locally and one test is failing. I was confused at this originally, it seems like the tests fail as there is error output but that's just misleading.
PS: I'd be willing to skip that test if that file doesn't exist in $path ;-)

**Comment by obbardc** [comment]
@evelikov This needs rebasing.

**Comment by obbardc** [comment]
@evelikov I have rebased your branch and solved the conflicts. I cannot access your fork. Here's the git log:

```
$ git push evelikov HEAD --force
Enumerating objects: 352, done.
Counting objects: 100% (309/309), done.
Delta compression using up to 16 threads
Compressing objects: 100% (200/200), done.
Writing objects: 100% (252/252), 47.69 KiB | 3.97 MiB/s, done.
Total 252 (delta 134), reused 71 (delta 30), pack-reused 0
remote: Resolving deltas: 100% (134/134), completed with 28 local objects.
To github.com:evelikov/debos.git
 ! [remote rejected] HEAD -> evelikov/remove-logStart (permission denied)
error: failed to push some refs to 'github.com:evelikov/debos.git'
```

**Comment by obbardc** [comment]
FTR, `git push -u evelikov HEAD:remove-logStart --force` worked.

**Pull request merged**
